### ISSUE TITLE
 resource "google_iap_client" "default" precondition failed

### ISSUE THREAD

**Issue by eahrend** [hashicorp/terraform-provider-google#8843]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Affected Resource(s)

* google_iap_client

### Terraform Configuration Files

```tf
resource "google_iap_client" "default" {
  display_name = var.iap_display_name
  brand        = "projects/###/brands/###"
}

```

### Debug Output
https://gist.github.com/eahrend/7a73cc0c4c3ac6ea7f80417ff3b66226

### Expected Behavior
Create the client using the brand id

### Actual Behavior
Gave a 400 bad request

### Steps to Reproduce

1. `terraform apply`

### Important Factoids
Using terragrunt as a wrapper for a bunch of different modules, but all of the data is hard coded to help debug.

### References

* #7334
Checked this github issue to see if there was a formatting issue, but my brand id was the same as the other ones, so it looks like it's fine.
* https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/iap_client
Used this doc to give me input on how to use the resource.

**Comment by slevenick** [comment]
Huh, I can't see why that would fail.

I'm able to successfully create the IAP client with the following request which looks nearly identical to yours:
```
---[ REQUEST ]---------------------------------------
POST /v1/projects/###/brands/###/identityAwareProxyClients?alt=json HTTP/1.1
Host: iap.googleapis.com
User-Agent: Terraform/0.14.8 (+https://www.terraform.io) Terraform-Plugin-SDK/2.5.0 terraform-provider-google-beta/dev
Content-Length: 30
Content-Type: application/json
Accept-Encoding: gzip

{
 "displayName": "Test Client"
}
```

Are you using the project number or the ID? It should be the number, which is what happens automatically when using `brand        =  google_iap_brand.project_brand.name` in the config. Since you are hardcoding it maybe you used a different format?

If that's not the case, can you try not including the number in your display name?

**Comment by eahrend** [comment]
Yeah, I tried it out using both the ID and the number. I couldn't get the `google_iap_brand` resource to work, since it was created by hand for testing purposes and I can't delete/recreate the brand. I managed to import it, but no outputs were able to come up since they apparently get rendered after an apply, and applying a `google_iap_brand` fails since one already exists. 

I actually got my project number and brand number from inspecting the network tab in the GCP console and looking at this tab via a request the browser made to cloudconsole-pa.clients6.google.com. 


```
resource "google_iap_client" "default" {
  display_name = var.iap_display_name
  brand        = "projects/424#####/brands/424#####"
}
```

**Comment by slevenick** [comment]
So you are able to import it, but not successfully run `terraform apply` after the import?

Can you create a client through the console for this brand? I'm guessing there is something that isn't working with the brand setup but that API error is very unhelpful

**Comment by codesuki** [comment]
We are running into the same issue. I just built a terraform module to set up IAP. Tested in a separate project, worked perfectly. Now trying to apply in our main project and we get ```Error: Error creating Client: googleapi: Error 400: Precondition check failed.```

Did you find out anything new @eahrend ?

**Comment by codesuki** [comment]
My co-worker @dtan4 found out that our brand was 'external' and the Google API only lets you create private IAP clients.

**Comment by eahrend** [comment]
@codesuki Interesting, yeah I haven't found anything in regards to that, but it seems like this is a breaking bug if we want external brands. Appreciate it.

**Comment by slevenick** [comment]
Interesting, so the API only works on internal brands, not external ones? The IAP API is pretty restrictive compared to what actions you can take in the console, I believe this is intentional, but that error message is really not ideal

**Comment by cathex-matt** [comment]
I experience the same issue right now but with an internal brand.

**Comment by cathex-matt** [comment]
> Yeah, I tried it out using both the ID and the number. I couldn't get the `google_iap_brand` resource to work, since it was created by hand for testing purposes and I can't delete/recreate the brand. I managed to import it, but no outputs were able to come up since they apparently get rendered after an apply, and applying a `google_iap_brand` fails since one already exists.
> 
> I actually got my project number and brand number from inspecting the network tab in the GCP console and looking at this tab via a request the browser made to cloudconsole-pa.clients6.google.com.
> 
> ```
> resource "google_iap_client" "default" {
>   display_name = var.iap_display_name
>   brand        = "projects/424#####/brands/424#####"
> }
> ```

I experience the exact same issue in the same use case right now, with an internal brand.

**Comment by cathex-matt** [comment]
1. the IAP client can't be created if the brand is external. And the type can't be changed programatically, it has to be done on GUI prior to apply your TF code. The 
2. the IAP brand once created can't be deleted nore updated
3. even when importing successfully the brand in the TF state, the apply keeps failing with `Error creating Brand: googleapi: Error 409: Requested entity already exists`

So here is my question now, what's the benefit of documenting an import [here](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/iap_brand#import) if it does not work.
And it's not documented anywhere how to handle this API behavior.

**Comment by mservidio** [comment]
> 1. the IAP client can't be created if the brand is external. And the type can't be changed programatically, it has to be done on GUI prior to apply your TF code.
> 2. the IAP brand once created can't be deleted nore updated (as documented by the Google Cloud platform)
> 3. BUT, even when importing successfully the brand in the TF state, the apply keeps failing with `Error creating Brand: googleapi: Error 409: Requested entity already exists`
> 
> So what is the benefit of documenting an import [here](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/iap_brand#import) if it does not work. And it's not documented anywhere how to handle this API behavior.

I'm running into the same exact issue too. I flipped it back to internal, but still get this 409 error.

**Comment by cathex-matt** [comment]
wow, good catch @mservidio
And even works after importing the `google_iap_brand`. At least that's what I thought, until I get stuck on a `Error: Error creating Client: googleapi: Error 400: Precondition check failed.`... 😞

So maybe my first deployment failed in the middle. I had to finally manually create the brand on console.
And then it worked creating the oauth client as defined by my template.

I want to highlight I'm using the TF provider 4.14.0 GA release. What would it be if trying the beta one...

**Comment by pv-aiuta-com** [comment]
Still does not work.
It worked when the IAP brand was internal.
Later I had to change it to external to use Identity platform for external users.
And now I can make new oauth clients via Web Console UI and import them to Terraform. 
But I can not create oauth client using gcloud CLI or Terraform

**Comment by enchorb** [comment]
I can't even import get the error `Error when reading or editing IapClient googleapi: Error 403: The caller does not have permission` - even though I have org admin and admin permissions...
